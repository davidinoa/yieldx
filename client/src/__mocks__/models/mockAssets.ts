import { Asset } from "@bondhouse/rover-universe";
import USD from "./assets/USD";
import muni from "./assets/muni";
import fund from "./assets/fund";

const mockAssets: Record<string, Asset> = {
  USD,
  muni
};

export const mockMultiAssets: Asset[] = [USD, muni, fund];

export default mockAssets;
