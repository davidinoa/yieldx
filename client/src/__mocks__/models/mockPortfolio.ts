import {
  Portfolio,
  PortfolioStatus,
  Position
} from "@bondhouse/position-management";
import { Asset } from "@bondhouse/rover-universe";
import muni from "./assets/muni";
import USD from "./assets/USD";

export function position(asset: Asset, quantity: number): Position {
  const { id } = asset;
  return {
    id,
    assetId: id,
    quantity,
    portfolioId: "axp"
  };
}

const mockPortfolio: Portfolio = {
  id: "axp",
  name: "",
  currency: "USD",
  status: PortfolioStatus.PENDING,
  positions: [position(muni, 5000), position(USD, 54450)],
  createdAt: "2020-06-01T17:01:36.467Z",
  metadata: {
    app: "Custom"
  }
};

export default mockPortfolio;
