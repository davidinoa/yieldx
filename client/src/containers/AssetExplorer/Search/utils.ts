interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
