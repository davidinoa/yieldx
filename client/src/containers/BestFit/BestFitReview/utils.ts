import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
