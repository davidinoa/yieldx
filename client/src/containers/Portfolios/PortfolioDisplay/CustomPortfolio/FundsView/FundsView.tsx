import React, { useState } from "react";
import { Col, Row } from "react-bootstrap";
import { ColumnApi, GridApi } from "@ag-grid-community/core";
import { useRecoilValue } from "recoil";
import { Card } from "elements/Card/Card";
import Decimal from "util/Decimal";
import { CardValue } from "elements/Card/Card.styles";
import { Spinner } from "elements/Spinner/Spinner";
import { ToolTipContent } from "assets/content";
import YieldXPortfolio from "models/YieldXPortfolio";
import exportDataAsCsv from "util/exportDataAsCsv";
import { getPositionData } from "recoil-state/getPositionData";
import { analyzeSleeve } from "recoil-state/portfoliosApi";
import PositionGridAll from "containers/Portfolios/PortfolioDisplay/PositionGrid/PositionGridAll";
import PortfolioSection from "containers/Portfolios/PortfolioDisplay/PortfolioSection/PortfolioSection";
import { PortfolioWidget } from "containers/Portfolios/PortfolioDisplay/PortfolioDisplay.styles";
import PortfolioFootNote from "containers/Portfolios/PortfolioDisplay/PortfolioFootNote/PortfolioFootNote";
import RiskCard from "containers/Portfolios/PortfolioDisplay/RiskCard/RiskCard";
import PortfolioGrowthChart from "containers/Portfolios/PortfolioDisplay/PortfolioGrowthChart/PortfolioGrowthChart";
import YieldContribution from "containers/Portfolios/PortfolioDisplay/Charts/YieldContribution/YieldContribution";
import SectorAllocation from "containers/Portfolios/PortfolioDisplay/Charts/SectorAllocation/SectorAllocation";
import { calculateYieldContributions } from "util/calculateYieldContributions";
import { FundPositions } from "../CustomPortfolio.styles";

interface Props {
  portfolio: YieldXPortfolio;
  version?: "CURRENT" | "ORIGINAL";
  editPositions: () => void;
}

function FundsView({ portfolio, version, editPositions }: Props) {
  const analysis = useRecoilValue(
    analyzeSleeve({
      drawDown: true,
      posData: getPositionData(portfolio)
    })
  );

  // @ts-expect-error
  const annual = analysis?.yield || 1;
  const [gridApi, setGridApi] = useState<GridApi>();
  const [columnApi, setColumnApi] = useState<ColumnApi>();
  const risk = analysis?.risk ?? 0;
  const maxDrawdown = analysis?.estimatedPerformance
    ? analysis.estimatedPerformance.estimatedMaximumDrawdown.value * 100
    : 0;
  const contribution = calculateYieldContributions(portfolio.positions, annual);
  return (
    <>
      <PortfolioSection title="Summary" version={version}>
        <Row>
          <Col xs={6}>
            <PortfolioWidget>
              <Row>
                <Card title="Yield" tooltipContent={ToolTipContent.Yield}>
                  <CardValue>
                    {annual ? (
                      <>
                        {Decimal.format(annual)}
                        <sup>%</sup>
                      </>
                    ) : (
                      <Spinner />
                    )}
                  </CardValue>
                </Card>
              </Row>
            </PortfolioWidget>
          </Col>
          <Col xs={6}>
            <PortfolioWidget>
              <Row>
                <Card
                  title="Distribution Income"
                  tooltipContent={ToolTipContent.DistributionIncome}
                >
                  <CardValue>
                    {annual ? (
                      <>
                        $
                        {Decimal.format(
                          (annual / 100 / 12) * (analysis.netAssetValue ?? 0)
                        )}
                        <sup>/m</sup>
                      </>
                    ) : (
                      <Spinner />
                    )}
                  </CardValue>
                </Card>
              </Row>
            </PortfolioWidget>
          </Col>
          <Col xs={6}>
            <PortfolioWidget>
              <Row>
                <Card
                  title="Expense Ratio"
                  tooltipContent={ToolTipContent.ExpenseRatio}
                >
                  <CardValue>
                    {analysis.expenseRatio ? (
                      <>
                        {Decimal.format(analysis.expenseRatio)}
                        <sup>%</sup>
                      </>
                    ) : (
                      <Spinner />
                    )}
                  </CardValue>
                </Card>
              </Row>
            </PortfolioWidget>
          </Col>
          <Col xs={6}>
            <PortfolioWidget>
              <Row>
                <RiskCard risk={risk} maxDrawdown={maxDrawdown} />
              </Row>
            </PortfolioWidget>
          </Col>
        </Row>
      </PortfolioSection>

      <PortfolioSection title="Analytics">
        <Row>
          <Col xs={12}>
            <SectorAllocation
              type="funds"
              positions={portfolio.positions}
              totalMarketValue={portfolio.totalMarketValue}
              cashMarketValue={portfolio.cashMarketValue}
              tooltipContent={ToolTipContent.SectorAllocation}
            />
          </Col>
          <Col xs={12}>
            <YieldContribution
              contribution={contribution}
              tooltipContent={ToolTipContent.YieldContribution.Funds}
            />
          </Col>
        </Row>
      </PortfolioSection>
      {analysis.wealthProjections && analysis.netAssetValue && (
        <PortfolioGrowthChart
          nav={analysis.netAssetValue}
          wealthProjections={analysis.wealthProjections}
          tooltipContent={ToolTipContent.PortfolioGrowth}
        />
      )}
      <FundPositions>
        <PortfolioSection
          title="Positions"
          exportPositions={() => exportDataAsCsv(gridApi, columnApi)}
          tooltipContent={ToolTipContent.Positions.Default}
          onEditPositions={editPositions}
        >
          <Row>
            <PositionGridAll
              sleeve="funds"
              cashAvailable={portfolio.cashMarketValue}
              portfolio={portfolio}
              setColumnApi={setColumnApi}
              setGridApi={setGridApi}
            />
          </Row>
          <Row>
            <PortfolioFootNote />
          </Row>
        </PortfolioSection>
      </FundPositions>
    </>
  );
}

export default FundsView;
