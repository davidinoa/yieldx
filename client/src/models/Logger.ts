/* eslint-disable @typescript-eslint/no-explicit-any */
import debug from "debug";

const colors: Record<string, string> = {
  debug: "#54c1e9",
  info: "#8cd10d",
  warn: "#fac848",
  error: "#f95c5a"
};

class Logger {
  private readonly enabled: boolean;

  constructor() {
    this.enabled = process.env.NODE_ENV !== "production";
  }

  debug(...args: any[]) {
    if (this.enabled) {
      Logger.generateMessage("debug", args);
    }
  }

  info(...args: any[]) {
    if (this.enabled) {
      Logger.generateMessage("info", args);
    }
  }

  warn(...args: any[]) {
    if (this.enabled) {
      Logger.generateMessage("warn", args);
    }
  }

  error(...args: any[]) {
    if (this.enabled) {
      Logger.generateMessage("error", args);
    }
  }

  private static generateMessage(level: string, args: any[]) {
    const createDebug = debug(`yieldx:${level}`);
    createDebug.color = colors[level];
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    createDebug(...args);
  }
}

const logger = new Logger();

export default logger;
