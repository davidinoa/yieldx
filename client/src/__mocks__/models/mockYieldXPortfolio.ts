import { PortfolioType } from "@bondhouse/income-portfolios";
import YieldXPortfolio from "models/YieldXPortfolio";
import mockPortfolio, { position } from "./mockPortfolio";
import mockAssets from "./mockAssets";
import fund from "./assets/fund";

export const mockMultiAssetPortfolio = new YieldXPortfolio(
  {
    ...mockPortfolio,
    positions: [...mockPortfolio.positions, position(fund, 200)]
  },
  Object.values(mockAssets).concat(fund),
  { portfolioType: PortfolioType.LIVE }
);

export default new YieldXPortfolio(mockPortfolio, Object.values(mockAssets));
