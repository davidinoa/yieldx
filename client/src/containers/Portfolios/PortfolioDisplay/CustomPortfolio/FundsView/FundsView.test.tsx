import React from "react";
import userEvent from "@testing-library/user-event";
import FundsView from "./FundsView";
import { render, screen } from "../../../../../test/utils";
import { mockMultiAssetPortfolio } from "../../../../../__mocks__/models/mockYieldXPortfolio";

describe("FundsView", () => {
  it("Should match snapshot", () => {
    const { asFragment } = render(
      <FundsView
        editPositions={jest.fn}
        portfolio={mockMultiAssetPortfolio.getFundSleeve()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders funds view", async () => {
    render(
      <FundsView
        editPositions={jest.fn}
        portfolio={mockMultiAssetPortfolio.getFundSleeve()}
      />
    );
    const editPositionButton = await screen.findByRole("button", {
      name: /edit positions/i
    });
    userEvent.click(editPositionButton);
  });
});
