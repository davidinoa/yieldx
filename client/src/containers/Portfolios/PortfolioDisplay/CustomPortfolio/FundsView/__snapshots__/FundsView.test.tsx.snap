// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FundsView Should match snapshot 1`] = `
<DocumentFragment>
  .c2 {
  margin: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  font-size: 0.5rem;
  color: var(--lightest);
}

.c1 {
  height: 100%;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  width: 100%;
  height: 100%;
}

<div
    class="c0"
    height="100%"
  >
    <div
      class="c1"
    >
      <div
        class="c2 spinner-border"
        data-testid="spinner"
      />
    </div>
  </div>
</DocumentFragment>
`;
